[gitops] support multiple SoftwarePackageSpecs in one yaml file

fuhry opened the pull request:
The list of software configured via GitOps might be dynamic, for example if the GitOps pipeline builds and pushes packages just before fleet gitops is called, or if partial package lists are shared between teams.

In these cases, it's useful to be able to include one file containing multiple packages. This PR refactors `parseSoftware` to support that use case.

This remains backward compatible with the old style of using one package per included file. This is verified by a new test, `TestSoftwarePackagesUnmarshalMulti`.

# Checklist for submitter

If some of the following don't apply, delete the relevant line.

<!-- Note that API documentation changes are now addressed by the product design team. -->

- [x] Changes file added for user-visible changes in `changes/`, `orbit/changes/` or `ee/fleetd-chrome/changes`.
  See [Changes files](https://github.com/fleetdm/fleet/blob/main/docs/Contributing/guides/committing-changes.md#changes-files) for more information.
- [x] Input data is properly validated, `SELECT *` is avoided, SQL injection is prevented (using placeholders for values in statements)
- [x] Added/updated automated tests

lukeheath commented:
@fuhry Thanks for your contribution! This looks really cool. Per our [community contribution process](https://fleetdm.com/handbook/engineering#review-a-community-pull-request), I'm assigning to the Engineering Manager for our Software product group (@mostlikelee) who will serve as your primary point of contact for this PR.

marko-lisica commented:
Hey @fuhry, thanks for your contribution. I created an issue to track this feature: https://github.com/fleetdm/fleet/issues/30849 

According to our process, the product team is responsible for reviewing all changes to the product. We think that this is really useful feature and we decided to prioritize this. We'll review it ASAP and inform you if any changes are necessary.

cc @mostlikelee

iansltx commented:
@fuhry As a quick update here, we just talked through this work our User story review ritual and are working through a couple nuances:

## Interaction with policy automations

Right now, the normal practice for specifying policy automations involving software is to specify the software YAML path. This stops working when a package path includes multiple packages. Workaround is to use `hash_sha256`, but that only works if a hash exists.

So, to get this mergeable, we'll at least need to hard-fail if a policy automation refers to a package YAML file containing multiple packages (haven't reviewed in depth; maybe you've already done this?). Bonus points for being able to specify `url:` instead of `hash_sha256:` in a policy automation spec, which would solve the other issue mentioned above, but we're fine deferring that work 'til later.

## Interaction with macOS setup experience

We're deprecating specifying software YAML files as part of the macOS setup experience item in favor of a flag on the software itself as part of #30095, but for multi-package YAML files we need to either ensure providing one serves as "select all of these" or fails with a not-supported error. We'll want to do one or the other as part of this PR.

## Interaction with upcoming software attribute location changes

We're moving where some keys live, from the software YAML file up to the team level, in #30095. That interacts with your changes here. Our current thought is that, for the keys that we're moving (label membership, self-service, categories), if you supplied those keys alongside a multi-package YAML file, that would effectively bulk apply that config for that team. No action required on your part, and the current plan is to merge your PR ahead of doing that work, but figured we'd give a heads-up.

---

If you don't have time to make the first or second tweaks for the above list, ping here and we'll leave dev capacity on our side to get those squared away (in addition to reviewing your work). Thanks again for your work on this!

fuhry commented on docs/Configuration/yaml-files.md:
````diff
@@ -501,6 +502,25 @@ You can get an output similar to that below when `fleetctl upload-software` succ
 - hash_sha256: fd22528a87f3cfdb81aca981953aa5c8d7084581b9209bb69abf69c09a0afaaf
 ```
 
+##### Multiple packages in a single included file
````
Hi Ian, sounds good, I'll get this split off for you momentarily.

fuhry committed:
[gitops] support multiple SoftwarePackageSpecs in one yaml file

The list of software configured via GitOps might be dynamic, for example if the GitOps pipeline builds and pushes packages just before fleet gitops is called, or if partial package lists are shared between teams.

In these cases, it's useful to be able to include one file containing multiple packages. This PR refactors `parseSoftware` to support that use case.

This remains backward compatible with the old style of using one package per included file. This is verified by a new test, `TestSoftwarePackagesUnmarshalMulti`.

Signed-Off-By: Dan Fuhry <dan@fuhry.com>

fuhry commented:
> So, to get this mergeable, we'll at least need to hard-fail if a policy automation refers to a package YAML file containing multiple packages (haven't reviewed in depth; maybe you've already done this?).

I did not explicitly do this, will work on this tomorrow for you.

> Bonus points for being able to specify `url:` instead of `hash_sha256:` in a policy automation spec, which would solve the other issue mentioned above, but we're fine deferring that work 'til later.

Will do if it doesn't blow up the scope/complexity of this PR.

> ## Interaction with macOS setup experience
> 
> We're deprecating specifying software YAML files as part of the macOS setup experience item in favor of a flag on the software itself as part of #30095, but for multi-package YAML files we need to either ensure providing one serves as "select all of these" or fails with a not-supported error. We'll want to do one or the other as part of this PR.

Ack.

> ## Interaction with upcoming software attribute location changes
> 
> We're moving where some keys live, from the software YAML file up to the team level, in #30095. That interacts with your changes here. Our current thought is that, for the keys that we're moving (label membership, self-service, categories), if you supplied those keys alongside a multi-package YAML file, that would effectively bulk apply that config for that team. No action required on your part, and the current plan is to merge your PR ahead of doing that work, but figured we'd give a heads-up.

Ack.

noahtalerman commented:
> Bonus points for being able to specify url: instead of hash_sha256: in a policy automation spec, which would solve the other issue mentioned above, but we're fine deferring that work 'til later.

@fuhry just a heads up that we decided to not solve this now. Fleet might come back to adding a `url` later but we don't want to do it in this iteration to keep the scope small.

fuhry commented:
@iansltx Thanks! I'll rebase on to any PRs related to #30095 today and resolve any merge conflicts so you aren't blocked.

iansltx commented:
@fuhry I wound up needing to re-roll your work on top of other changes, as #32503. That's been merged and will go out in Fleet 4.74. You'll get a shout-out in today's sprint demos for this work; thanks!
